Add BUILD_URL in StepMetadata.Env()

analytically opened the pull request:
This commit:
- Adds a BUILD_URL env parameter to step metadata
- Adds proper handling for one-off jobs with a different URL pattern

Was also requested here: https://github.com/concourse/concourse/issues/1588

analytically commented:
Also discussed at length here: https://github.com/concourse/concourse/pull/2747

kcbimonte commented:
So, we developed a resource internally that allows for Concourse to "run" external pipelines and report them as such to GitLab. GitLab has a limit of 255 characters for Target URLs ([Commit Status API](https://docs.gitlab.com/api/commits/#set-the-pipeline-status-of-a-commit)), so we instead went with using the Build ID path style instead (e.g. https://ci.concourse-ci.org/builds/411969621).

Since certain tools have path length limitations on where we'd realistically use this BUILD_URL variable, does it make more sense to instead just compose the BUILD_URL based off the `$ATC_EXTERNAL_URL/builds/$BUILD_ID` so that we can stay within the sensible path limits while also providing a ready-made URL?

analytically committed:
Add BUILD_URL in StepMetadata.Env()

This commit:
- Adds a BUILD_URL env parameter to step metadata
- Adds proper handling for one-off jobs with a different URL pattern
- Maintains detailed URL format when possible for better readability
- Ensures compatibility with GitLab and other tools with path length limitations

Was also requested here: #1588

Signed-off-by: Mathias Bogaert <mathias.bogaert@gmail.com>

analytically commented:
@kcbimonte I've added a test for > 255 char length limit and used the shorter version when that's the case

taylorsilva reviewed (changes requested):
Would be nice to add an acceptance level test as well: https://github.com/concourse/concourse/pull/2747#issuecomment-439919525
